src/ir.h

```diff
@@ -188,6 +188,13 @@ struct FuncSignature {
   TypeVector param_types;
   TypeVector result_types;
 
+  // Some types can have names, for example (ref $foo) has type $foo.
+  // So to use this type we need to translate its name into
+  // a proper index from the module type section.
+  // This is the mapping from parameter/result index to its name.
+  std::unordered_map<uint32_t, std::string> param_type_names;
+  std::unordered_map<uint32_t, std::string> result_type_names;
```

Comment:
I can't tell from this comment why the signature needs to store type names, since there is no signature involved in `(ref $foo)`. Shouldn't all type names already be stored somewhere else already?

Reply:
actually we should store type names because at the moment of parsing we may not know all the types.
We will resolve these type names into indices after the parsing completes and then we don't need to use these maps at all

Reply:
> Is there a precedent for that type of thing in the codebase already?

Sure, for example names of locals handled the same way:
```
(module
  (func (param $x i32) (result i32)
    (local.get $x)
  )
)
```
In the example above we have `$x` name for the first parameter and in current design we store these mapping `$x` -> `0` here: https://github.com/WebAssembly/wabt/blob/d61a78988001cdc6d364d91cfb9b2088b9ae2798/src/ir.h#L767

We set it up inside
```
Result WastParser::ParseFuncSignature(FuncSignature* sig,
                                      BindingHash* param_bindings)
```
And this information also actually needed only for parsing, but here we are.

BTW, the ideal solution is to change Type representation and make it like `struct Var` to handle string names and indices but because of PODness of `Type` we can't do much without some refactoring.